### ISSUE TITLE
TestLucene90DocValuesFormat fails with ArrayIndexOutOfBoundsException

### ISSUE THREAD

**Issue by ChrisHegarty** [apache/lucene#13805]
```
ERROR: The following test(s) have failed:
  - org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.testSparseDocValuesVsStoredFields (:lucene:core)
    Test output: /opt/buildkite-agent/builds/bk-agent-prod-gcp-1726674638633683811/elastic/apache-lucene-nightly/lucene/core/build/test-results/test/outputs/OUTPUT-org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.txt
    Reproduce with: gradlew :lucene:core:test --tests "org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.testSparseDocValuesVsStoredFields" -Ptests.jvms=12 -Ptests.jvmargs= -Ptests.seed=175AB2293A24B66E -Ptests.nightly=true -Ptests.gui=true -Ptests.file.encoding=UTF-8 -Ptests.vectorsize=128 -Ptests.forceintegervectors=true
  - org.apache.lucene.backward_codecs.lucene80.TestBestSpeedLucene80DocValuesFormat.testSparseDocValuesVsStoredFields (:lucene:backward-codecs)
    Test output: /opt/buildkite-agent/builds/bk-agent-prod-gcp-1726674638633683811/elastic/apache-lucene-nightly/lucene/backward-codecs/build/test-results/test/outputs/OUTPUT-org.apache.lucene.backward_codecs.lucene80.TestBestSpeedLucene80DocValuesFormat.txt
    Reproduce with: gradlew :lucene:backward-codecs:test --tests "org.apache.lucene.backward_codecs.lucene80.TestBestSpeedLucene80DocValuesFormat.testSparseDocValuesVsStoredFields" -Ptests.jvms=12 -Ptests.jvmargs= -Ptests.seed=175AB2293A24B66E -Ptests.nightly=true -Ptests.gui=true -Ptests.file.encoding=UTF-8 -Ptests.vectorsize=128 -Ptests.forceintegervectors=true
```

```
 >     java.lang.ArrayIndexOutOfBoundsException: Index 3 out of bounds for length 3
   >         at __randomizedtesting.SeedInfo.seed([175AB2293A24B66E:43816E3AC0A89021]:0)
   >         at org.apache.lucene.util.packed.Packed64.get(Packed64.java:80)
   >         at org.apache.lucene.index.OrdinalMap$1.get(OrdinalMap.java:379)
   >         at org.apache.lucene.codecs.DocValuesConsumer$7$1.nextOrd(DocValuesConsumer.java:946)
   >         at org.apache.lucene.codecs.lucene90.Lucene90DocValuesConsumer$4$1.nextDoc(Lucene90DocValuesConsumer.java:808)
   >         at org.apache.lucene.codecs.lucene90.Lucene90DocValuesConsumer.writeValues(Lucene90DocValuesConsumer.java:201)
   >         at org.apache.lucene.codecs.lucene90.Lucene90DocValuesConsumer.doAddSortedNumericField(Lucene90DocValuesConsumer.java:705)
   >         at org.apache.lucene.codecs.lucene90.Lucene90DocValuesConsumer.addSortedSetField(Lucene90DocValuesConsumer.java:770)
   >         at org.apache.lucene.codecs.DocValuesConsumer.mergeSortedSetField(DocValuesConsumer.java:853)
   >         at org.apache.lucene.codecs.DocValuesConsumer.merge(DocValuesConsumer.java:148)
   >         at org.apache.lucene.codecs.perfield.PerFieldDocValuesFormat$FieldsWriter.merge(PerFieldDocValuesFormat.java:152)
   >         at org.apache.lucene.index.SegmentMerger.mergeDocValues(SegmentMerger.java:188)
   >         at org.apache.lucene.index.SegmentMerger.mergeWithLogging(SegmentMerger.java:314)
   >         at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:149)
   >         at org.apache.lucene.index.IndexWriter.mergeMiddle(IndexWriter.java:5292)
   >         at org.apache.lucene.index.IndexWriter.merge(IndexWriter.java:4758)
   >         at org.apache.lucene.index.IndexWriter$IndexWriterMergeSource.merge(IndexWriter.java:6581)
   >         at org.apache.lucene.index.SerialMergeScheduler.merge(SerialMergeScheduler.java:38)
   >         at org.apache.lucene.index.IndexWriter.executeMerge(IndexWriter.java:2327)sFormat
   >         at org.apache.lucene.index.IndexWriter.maybeMerge(IndexWriter.java:2322)
   >         at org.apache.lucene.index.IndexWriter.processEvents(IndexWriter.java:6033)
   >         at org.apache.lucene.index.IndexWriter.maybeProcessEvents(IndexWriter.java:6023)
   >         at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1562)
   >         at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1847)
   >         at org.apache.lucene.index.IndexWriter.addDocument(IndexWriter.java:1487)
   >         at org.apache.lucene.tests.index.RandomIndexWriter.addDocument(RandomIndexWriter.java:224)
   >         at org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.doTestSparseDocValuesVsStoredFields(TestLucene90DocValuesFormat.java:215)
   >         at org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.testSparseDocValuesVsStoredFields(TestLucene90DocValuesFormat.java:169)
   >         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
   >         at java.base/java.lang.reflect.Method.invoke(Method.java:580)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
   >         at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
   >         at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
   >         at org.junit.rules.RunRules.evaluate(RunRules.java:20)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
   >         at org.junit.rules.RunRules.evaluate(RunRules.java:20)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
   >         at java.base/java.lang.Thread.run(Thread.java:1583)
```

**Comment by benwtrent** [comment]
Git bisect puts the blame at: 6634b41f42f4e2802048d1e4750e1ce1202652c5

https://github.com/apache/lucene/pull/13686

**Comment by benwtrent** [comment]
git bisect might be lying, I don't see how that PR could cause this failure :(

**Comment by iverase** [comment]
Probably we are not remapping the field ordinal properly when merging segments.

**Comment by iverase** [comment]
See here: https://github.com/apache/lucene/blob/6d987e1ce1c3f3215633a979ce048829fe1bb6ed/lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90PointsWriter.java#L248

**Comment by jpountz** [comment]
Argh, I remember carefully checking whether this PR could cause issues due to mismatched field infos, but apparently I missed something.

**Comment by rmuir** [comment]
Can we just revert the change for now? it does two things at once... one of those is using field.number instead of field.name which is historically unsafe: it was always the big risk of bulk merge.

It can't be done like this in all situations, and doing it across versions like this is especially YOLO and asking for corruption IMO. This is why stored fields never bulk merge across versions: https://github.com/apache/lucene/blob/e4ac57746eb86846b3a53944c14e09873f793ff1/lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsWriter.java#L684

**Comment by ChrisHegarty** [comment]
I'm going to try reverting https://github.com/apache/lucene/commit/6634b41f42f4e2802048d1e4750e1ce1202652c5.

**Comment by benwtrent** [comment]
@ChrisHegarty this makes be worried about all the other field number switch with field name things as well. 

I am wondering if we should revert all of them, there are multiple PRs.

**Comment by ChrisHegarty** [comment]
The revert fixes the failures we see here and the other related test failures, seen in #13807 #13808.

**Comment by rmuir** [comment]
sounds like the safe bet to backout any changes messing around with fieldinfos on merge.

Sorry for the short explanation, there is a long history of super-sneaky corruption bugs like this. always happening on some corner-case such as `addIndexes(reader)` or across different versions, or something like that. When they happen on merge it makes debugging them especially difficult. Mixing up data across fields because of field numbers happened more than once.

This is why, if you look at bulk merge code, you see crazy sysprop escape hatches and stuffl like that: https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsWriter.java#L492-L508

**Comment by rmuir** [comment]
we didn't even have bulk merge at all in lucene for a couple years at all because of field-number bugs like this. got bit too many times.

**Comment by ChrisHegarty** [comment]
I filed a meta issue to better track the reverts, #13809

**Comment by jpountz** [comment]
I don't mind reverting but I would also like to fix the root cause as this change only exposed an existing bug: someone is calling a doc-values producer with the wrong FieldInfo object.

**Comment by rmuir** [comment]
yeah it would be best to improve the tests: it is not good that it took this test, run many many times, to find it.

**Comment by jpountz** [comment]
I found the bug, it's the slow composite reader wrapper which is at fault here. I'll look into improving tests to detect such issues.

Separately, we may want to consider changing the DocValuesProducer API to take a String rather than a FieldInfo, like e.g. points, so that it is not tempted to trust the caller to resolve the FieldInfo object correctly.

**Comment by jpountz** [comment]
I found the actual root cause, it's here: https://github.com/apache/lucene/blob/e4ac57746eb86846b3a53944c14e09873f793ff1/lucene/core/src/java/org/apache/lucene/codecs/DocValuesConsumer.java#L616. The producer is called on `fieldInfo` instead of `readerFieldInfo` like other doc values types do. I'm working on tests that would have uncovered this problem.

**Comment by ChrisHegarty** [comment]
Ok, reverts are prepared. @jpountz you wanna fix (and not revert), or revert for now?

**Comment by jpountz** [comment]
Give me some time to see how the fix and tests look, and let's think about whether/what to revert later on? I expect to have something by end of day. @ChrisHegarty Feel free to cut the branch in the meantime, we can backport to the 9.12 branch if necessary?

**Comment by ChrisHegarty** [comment]
Let's postpone the 9_12 branch cut until tomorrow, pending on the outcome of this.

**Comment by bugmakerrrrrr** [comment]
> Separately, we may want to consider changing the DocValuesProducer API to take a String rather than a FieldInfo, like e.g. points, so that it is not tempted to trust the caller to resolve the FieldInfo object correctly.

@jpountz +1, We have encountered the [related issue in NormsProducer](https://github.com/apache/lucene/issues/13685), and I worked around it by resolving the field info inside the NormsProducer.

**Comment by jpountz** [comment]
I have a fix and tests that would have found the bug at #13812.
